File: EduSchedule-test/resource-service/src/main/java/cm/iusjc/resource/service/SalleMaterielService.java
package cm.iusjc.resource.service;

import cm.iusjc.resource.entity.*;
import cm.iusjc.resource.entity.EquipementReservation.StatutReservationEquipement;
import cm.iusjc.resource.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Service gérant la relation formelle Salle ↔ Matériel et l'allocation
 * automatique d'équipements selon le type de cours (TD, TP, CM, EXAM…).
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class SalleMaterielService {

    private final SalleMaterielRepository salleMaterielRepo;
    private final EquipementReservationRepository equipReservationRepo;
    private final SalleRepository salleRepo;
    private final MaterielRepository materielRepo;

    // ── Mapping type de cours → codes de types de matériels requis ────────────

    /**
     * Retourne les codes de types de matériels recommandés pour un type de cours.
     * Ces codes correspondent aux TypeMateriel.code en base.
     */
    public static List<String> getEquipementsRequisPourTypeCours(String typeCours) {
        if (typeCours == null) return List.of();
        return switch (typeCours.toUpperCase()) {
            case "TP", "PRACTICAL" -> List.of("ORDINATEUR", "PROJECTEUR", "TABLEAU_BLANC");
            case "TD", "SEMINAR"   -> List.of("TABLEAU_BLANC", "PROJECTEUR");
            case "CM", "CONFERENCE" -> List.of("PROJECTEUR", "SONO", "MICRO");
            case "EXAM"            -> List.of("TABLEAU_BLANC");
            default                -> List.of("TABLEAU_BLANC");
        };
    }

    // ── Inventaire salle ──────────────────────────────────────────────────────

    public List<SalleMateriel> getInventaireSalle(Long salleId) {
        return salleMaterielRepo.findBySalleId(salleId);
    }

    public List<SalleMateriel> getEquipementsDisponibles(Long salleId) {
        return salleMaterielRepo.findDisponiblesBySalleId(salleId);
    }

    @Transactional
    public SalleMateriel ajouterMaterielDansSalle(Long salleId, Long materielId,
                                                   int quantite, boolean requis, String notes) {
        Salle salle = salleRepo.findById(salleId)
                .orElseThrow(() -> new RuntimeException("Salle introuvable: " + salleId));
        Materiel materiel = materielRepo.findById(materielId)
                .orElseThrow(() -> new RuntimeException("Matériel introuvable: " + materielId));

        // Vérifier si la relation existe déjà
        Optional<SalleMateriel> existing = salleMaterielRepo.findBySalleIdAndMaterielId(salleId, materielId);
        if (existing.isPresent()) {
            SalleMateriel sm = existing.get();
            sm.setQuantiteTotale(sm.getQuantiteTotale() + quantite);
            sm.setQuantiteDisponible(sm.getQuantiteDisponible() + quantite);
            if (notes != null) sm.setNotes(notes);
            return salleMaterielRepo.save(sm);
        }

        SalleMateriel sm = new SalleMateriel();
        sm.setSalle(salle);
        sm.setMateriel(materiel);
        sm.setQuantiteTotale(quantite);
        sm.setQuantiteDisponible(quantite);
        sm.setQuantiteReservee(0);
        sm.setRequis(requis);
        sm.setNotes(notes);
        sm.setDateInstallation(LocalDateTime.now());
        return salleMaterielRepo.save(sm);
    }

    @Transactional
    public SalleMateriel mettreAJourQuantite(Long salleMaterielId, int nouvelleQuantiteTotale) {
        SalleMateriel sm = salleMaterielRepo.findById(salleMaterielId)
                .orElseThrow(() -> new RuntimeException("SalleMateriel introuvable: " + salleMaterielId));
        int delta = nouvelleQuantiteTotale - sm.getQuantiteTotale();
        sm.setQuantiteTotale(nouvelleQuantiteTotale);
        sm.setQuantiteDisponible(Math.max(0, sm.getQuantiteDisponible() + delta));
        return salleMaterielRepo.save(sm);
    }

    @Transactional
    public void retirerMaterielDeSalle(Long salleMaterielId) {
        salleMaterielRepo.deleteById(salleMaterielId);
    }

    // ── Allocation automatique selon type de cours ────────────────────────────

    /**
     * Vérifie si une salle dispose des équipements requis pour un type de cours
     * sur une plage horaire donnée.
     */
    public Map<String, Object> verifierDisponibiliteEquipements(Long salleId, String typeCours,
                                                                  LocalDateTime debut, LocalDateTime fin) {
        List<String> codesRequis = getEquipementsRequisPourTypeCours(typeCours);
        List<SalleMateriel> inventaire = salleMaterielRepo.findBySalleId(salleId);

        List<Map<String, Object>> details = new ArrayList<>();
        boolean toutDisponible = true;

        for (String code : codesRequis) {
            List<SalleMateriel> matching = inventaire.stream()
                    .filter(sm -> sm.getMateriel().getTypeMateriel() != null
                            && code.equalsIgnoreCase(sm.getMateriel().getTypeMateriel().getCode()))
                    .collect(Collectors.toList());

            boolean disponible = false;
            int quantiteDisponible = 0;

            for (SalleMateriel sm : matching) {
                // Vérifier les conflits sur la plage horaire
                List<EquipementReservation> conflicts = equipReservationRepo.findConflicts(sm.getId(), debut, fin);
                int quantiteReserveeConflict = conflicts.stream().mapToInt(EquipementReservation::getQuantite).sum();
                int dispo = sm.getQuantiteTotale() - quantiteReserveeConflict;
                if (dispo > 0) {
                    disponible = true;
                    quantiteDisponible += dispo;
                }
            }

            if (!disponible) toutDisponible = false;

            Map<String, Object> detail = new LinkedHashMap<>();
            detail.put("typeEquipement", code);
            detail.put("disponible", disponible);
            detail.put("quantiteDisponible", quantiteDisponible);
            detail.put("requis", true);
            details.add(detail);
        }

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("salleId", salleId);
        result.put("typeCours", typeCours);
        result.put("toutDisponible", toutDisponible);
        result.put("equipements", details);
        return result;
    }

    /**
     * Alloue automatiquement les équipements requis pour un type de cours.
     * Appelé lors de la confirmation d'une réservation.
     *
     * @return liste des EquipementReservation créées
     */
    @Transactional
    public List<EquipementReservation> allouerEquipements(Long reservationId, Long salleId,
                                                           String typeCours,
                                                           LocalDateTime debut, LocalDateTime fin) {
        List<String> codesRequis = getEquipementsRequisPourTypeCours(typeCours);
        List<SalleMateriel> inventaire = salleMaterielRepo.findBySalleId(salleId);
        List<EquipementReservation> allocations = new ArrayList<>();

        for (String code : codesRequis) {
            List<SalleMateriel> matching = inventaire.stream()
                    .filter(sm -> sm.getMateriel().getTypeMateriel() != null
                            && code.equalsIgnoreCase(sm.getMateriel().getTypeMateriel().getCode())
                            && sm.getMateriel().getEtat() != Materiel.EtatMateriel.EN_PANNE
                            && sm.getMateriel().getEtat() != Materiel.EtatMateriel.EN_MAINTENANCE)
                    .collect(Collectors.toList());

            for (SalleMateriel sm : matching) {
                List<EquipementReservation> conflicts = equipReservationRepo.findConflicts(sm.getId(), debut, fin);
                int quantiteReserveeConflict = conflicts.stream().mapToInt(EquipementReservation::getQuantite).sum();
                int dispo = sm.getQuantiteTotale() - quantiteReserveeConflict;

                if (dispo > 0) {
                    EquipementReservation er = new EquipementReservation();
                    er.setReservationId(reservationId);
                    er.setSalleMateriel(sm);
                    er.setQuantite(1);
                    er.setTypeCours(typeCours);
                    er.setDateDebut(debut);
                    er.setDateFin(fin);
                    er.setStatut(StatutReservationEquipement.ACTIVE);
                    er.setNotes("Allocation automatique pour " + typeCours);
                    allocations.add(equipReservationRepo.save(er));
                    log.info("Équipement alloué: {} (salle {}) pour réservation {} [{}]",
                            sm.getMateriel().getNom(), salleId, reservationId, typeCours);
                    break; // Un seul équipement par type suffit
                }
            }
        }

        log.info("{} équipement(s) alloué(s) pour réservation {} (salle {}, type {})",
                allocations.size(), reservationId, salleId, typeCours);
        return allocations;
    }

    /**
     * Libère tous les équipements alloués pour une réservation (annulation).
     */
    @Transactional
    public void libererEquipements(Long reservationId) {
        List<EquipementReservation> actives = equipReservationRepo.findActiveByReservationId(reservationId);
        actives.forEach(er -> {
            er.setStatut(StatutReservationEquipement.ANNULEE);
            equipReservationRepo.save(er);
        });
        log.info("{} équipement(s) libéré(s) pour réservation {}", actives.size(), reservationId);
    }

    /**
     * Marque les équipements d'une réservation comme terminés.
     */
    @Transactional
    public void terminerEquipements(Long reservationId) {
        List<EquipementReservation> actives = equipReservationRepo.findActiveByReservationId(reservationId);
        actives.forEach(er -> {
            er.setStatut(StatutReservationEquipement.TERMINEE);
            equipReservationRepo.save(er);
        });
    }

    // ── Statistiques ──────────────────────────────────────────────────────────

    public Map<String, Object> getStatistiques() {
        List<Object[]> statsByType = equipReservationRepo.getStatsByTypeCours();
        Map<String, Object> stats = new LinkedHashMap<>();
        stats.put("parTypeCours", statsByType.stream().map(row -> {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("typeCours", row[0]);
            m.put("nombreReservations", row[1]);
            m.put("quantiteTotale", row[2]);
            return m;
        }).collect(Collectors.toList()));
        stats.put("totalSalleMateriel", salleMaterielRepo.count());
        return stats;
    }
}

File: EduSchedule-test/frontend-angular/src/app/core/services/salle-materiel.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable, of } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { ApiService } from './api.service';

export interface SalleMateriel {
  id: number;
  salle: { id: number; code: string; name: string; type: string };
  materiel: {
    id: number; code: string; nom: string;
    typeMateriel?: { code: string; nom: string; icone: string; couleur: string };
    marque: string; modele: string; etat: string;
  };
  quantiteTotale: number;
  quantiteDisponible: number;
  quantiteReservee: number;
  requis: boolean;
  notes: string;
  dateInstallation: string;
}

export interface EquipementReservation {
  id: number;
  reservationId: number;
  salleMateriel: SalleMateriel;
  quantite: number;
  typeCours: string;
  dateDebut: string;
  dateFin: string;
  statut: 'ACTIVE' | 'TERMINEE' | 'ANNULEE';
  notes: string;
}

export interface DisponibiliteEquipement {
  salleId: number;
  typeCours: string;
  toutDisponible: boolean;
  equipements: {
    typeEquipement: string;
    disponible: boolean;
    quantiteDisponible: number;
    requis: boolean;
  }[];
}

interface ApiWrapped<T> { success: boolean; data: T; total?: number; }

@Injectable({ providedIn: 'root' })
export class SalleMaterielService {
  private api = inject(ApiService);

  // ── Inventaire par salle ──────────────────────────────────────────────────

  getInventaireSalle(salleId: number): Observable<SalleMateriel[]> {
    return this.api.get<ApiWrapped<SalleMateriel[]>>(`/v1/salle-materiels/salle/${salleId}`).pipe(
      map(r => r?.data ?? []),
      catchError(() => of([]))
    );
  }

  getEquipementsDisponibles(salleId: number): Observable<SalleMateriel[]> {
    return this.api.get<ApiWrapped<SalleMateriel[]>>(`/v1/salle-materiels/salle/${salleId}/disponibles`).pipe(
      map(r => r?.data ?? []),
      catchError(() => of([]))
    );
  }

  ajouterMaterielDansSalle(salleId: number, materielId: number, quantite: number,
                            requis: boolean, notes: string): Observable<SalleMateriel | null> {
    return this.api.post<ApiWrapped<SalleMateriel>>(
      `/v1/salle-materiels/salle/${salleId}/materiel/${materielId}`,
      { quantite, requis, notes }
    ).pipe(
      map(r => r?.data ?? null),
      catchError(() => of(null))
    );
  }

  mettreAJourQuantite(id: number, quantite: number): Observable<SalleMateriel | null> {
    return this.api.put<ApiWrapped<SalleMateriel>>(`/v1/salle-materiels/${id}/quantite`, { quantite }).pipe(
      map(r => r?.data ?? null),
      catchError(() => of(null))
    );
  }

  retirerMateriel(id: number): Observable<boolean> {
    return this.api.delete<any>(`/v1/salle-materiels/${id}`).pipe(
      map(() => true),
      catchError(() => of(false))
    );
  }

  // ── Disponibilité et allocation ───────────────────────────────────────────

  verifierDisponibilite(salleId: number, typeCours: string,
                         debut: string, fin: string): Observable<DisponibiliteEquipement> {
    const params = `?typeCours=${typeCours}&debut=${encodeURIComponent(debut)}&fin=${encodeURIComponent(fin)}`;
    return this.api.get<DisponibiliteEquipement>(`/v1/salle-materiels/salle/${salleId}/disponibilite${params}`).pipe(
      catchError(() => of({ salleId, typeCours, toutDisponible: false, equipements: [] }))
    );
  }

  getEquipementsRequis(typeCours: string): Observable<string[]> {
    return this.api.get<any>(`/v1/salle-materiels/equipements-requis?typeCours=${typeCours}`).pipe(
      map(r => r?.equipementsRequis ?? []),
      catchError(() => of([]))
    );
  }

  allouerEquipements(reservationId: number, salleId: number, typeCours: string,
                      dateDebut: string, dateFin: string): Observable<EquipementReservation[]> {
    return this.api.post<ApiWrapped<EquipementReservation[]>>('/v1/salle-materiels/allouer', {
      reservationId, salleId, typeCours, dateDebut, dateFin
    }).pipe(
      map(r => r?.data ?? []),
      catchError(() => of([]))
    );
  }

  libererEquipements(reservationId: number): Observable<boolean> {
    return this.api.post<any>(`/v1/salle-materiels/liberer/${reservationId}`, {}).pipe(
      map(() => true),
      catchError(() => of(false))
    );
  }

  // ── Réservations d'équipements ────────────────────────────────────────────

  getReservationsEquipements(reservationId: number): Observable<EquipementReservation[]> {
    return this.api.get<ApiWrapped<EquipementReservation[]>>(`/v1/salle-materiels/reservations/${reservationId}`).pipe(
      map(r => r?.data ?? []),
      catchError(() => of([]))
    );
  }

  // ── Statistiques ──────────────────────────────────────────────────────────

  getStatistiques(): Observable<any> {
    return this.api.get<any>('/v1/salle-materiels/statistiques').pipe(
      catchError(() => of(null))
    );
  }
}
